menoh/mkldnn_with_generic_fallback/backend/mkldnn/operator/batch_norm.hpp

```diff
@@ -0,0 +1,118 @@
+                auto spatial = attribute_int(node, "spatial");
+                if(!spatial) {
+                    throw std::runtime_error(
+                      "Non spacial BatchNormalization is not supported");
+                }
+
+                memory_cache& input_memory_cache =
+                  input_memory_cache_list.at(0);
+                auto input_dims = input_memory_cache.dims();
+                auto input_memory = input_memory_cache.get_data_memory();
+
+                auto mean_memory =
+                  get_memory(input_memory_cache_list.at(3),
+                             mkldnn::memory::format::x, primitives);
+                auto var_memory =
+                  get_memory(input_memory_cache_list.at(4),
+                             mkldnn::memory::format::x, primitives);
+
+                memory_cache& scale_memory_cache =
+                  input_memory_cache_list.at(1);
+                auto scale_dims = scale_memory_cache.dims();
+                memory_cache& b_memory_cache = input_memory_cache_list.at(2);
+                std::vector<int> weight_dims({2});
+                weight_dims.insert(weight_dims.end(), scale_dims.begin(),
+                                   scale_dims.end());
+                auto scale_memory = get_memory(scale_memory_cache, scale_dims,
+                                               mkldnn::memory::x, primitives);
+                auto b_memory = get_memory(b_memory_cache, scale_dims,
+                                           mkldnn::memory::x, primitives);
+                mkldnn::memory weight_memory({{{weight_dims},
+                                               extract_data_type(scale_memory),
+                                               mkldnn::memory::format::nc},
+                                              engine});
+                auto scale_total_size =
+                  std::accumulate(scale_dims.begin(), scale_dims.end(), 1,
+                                  std::multiplies<int>());
+                // TODO other data_type
+                if(extract_data_type(scale_memory) ==
+                   mkldnn::memory::data_type::f32) {
+                    std::copy(
+                      static_cast<float*>(scale_memory.get_data_handle()),
+                      static_cast<float*>(scale_memory.get_data_handle()) +
+                        scale_total_size,
+                      static_cast<float*>(weight_memory.get_data_handle()));
+                    std::copy(
+                      static_cast<float*>(b_memory.get_data_handle()),
+                      static_cast<float*>(b_memory.get_data_handle()) +
+                        scale_total_size,
+                      static_cast<float*>(weight_memory.get_data_handle()) +
+                        scale_total_size);
+                } else {
+                    throw std::runtime_error(
+                      "BatchNormalization with invalid data_type");
+                }
```

Comment:
It seems that the copy from `scale_memory`/`b_memory` to `weight_memory` is performed in the MKL-DNN computation graph construction time, not execution time. Isn't this problematic when `scale_memory`/`b_memory`  are not constant tensor?

Reply:
I see.